### DISCUSSION TITLE
Support Complex Number Type

### DISCUSSION THREAD

**Discussion by tigercosmos** [solvcon/modmesh#383] [Ideas]
add complex in `SimplArray`

**Comment by yungyuc** [comment]
@tigercosmos please provide more analyses

**Reply by tigercosmos** [reply]
sure, will add the details later. studying Numpy now.

**Reply by tigercosmos** [reply]
We can have a simple strategy.

- add a new type in SimpleArray
- a struct `Complex`, which includes two floating numbers.
- store the complex number continuously in the memory

following this concept, should not be hard to implement.

**Reply by yungyuc** [reply]
What is the memory layout of the complex-valued array?  There are prior arts and analyses should be made before implementation.

**Reply by tigercosmos** [reply]
I read the source code of numpy, and also I searched the C++ spec.
I think the most common and efficient way is to combine real and imaginary parts together.

for example, for a three-element complex64 array, the layout will be:
```
| R (4 bytes) | I (4 bytes) | R (4 bytes) | I (4 bytes) | R (4 bytes) | I (4 bytes) |
```

when we use the complex number, usually we will need both real and imaginary parts, so this layout ensures the memory efficiency.

**Reply by yungyuc** [reply]
It sounds like a plan to start with following [the `std::complex`](https://en.cppreference.com/w/cpp/numeric/complex) array layout, but we need a fuller analysis included in a specification with critical review.  Let us use this thread to collect references.  Could you please organize all prior arts for computer codes handling complex numbers?

**Reply by tigercosmos** [reply]
@yungyuc  I checked the implementation of Numpy, std::complex, Eigen, [FFTW](https://www.fftw.org/).

- Numpy
  - NumPy defines its own `numpy.complex64` and `numpy.complex128` types, which are Python objects
  - Complex arrays are stored as interleaved real and imaginary parts in contiguous memory blocks.
- std::complex
  - the complex is a struct containing two members: the real part and the imaginary part.
  - Complex arrays are stored as interleaved real and imaginary parts in contiguous memory blocks.
- Eigen
  - leverage the `std::complex`
  - Complex arrays are stored as interleaved real and imaginary parts in contiguous memory blocks.
- FFTW
  - use C99 complex type, which is two consecutive floating-point numbers
  - Complex arrays are stored as interleaved real and imaginary parts in contiguous memory blocks.

To sum up, other works may have different complex types, but they all store the complex arrays in an interleaved way. Therefore, we can define our complex type, which should include the real and imaginary parts, and store the complex numbers in an interleaved array.

**Reply by yungyuc** [reply]
This is a list good enough to serve as a fundamental understanding for starting to prototype the complex value support for simple array, but it can use more clarity.  A key point is to provide a reference and description to the interleaved storage style of the real and imaginary parts in a complex-valued array.

FFTW described the "[interleaved format](https://www.fftw.org/fftw3_doc/Interleaved-and-split-arrays.html)" as:
> In the interleaved format, you provide pointers to the real part of a complex number, and the imaginary part understood to be stored in the next memory location.

The other storage style for the complex-valued array in FFTW is the *split format*:
> The split format allows separate pointers to the real and imaginary parts of a complex array.

It helps to clarify that how the scalars `complex64` and `complex128` in Numpy are laid out in memory.  It is defined in [`npy_common.h`](https://github.com/numpy/numpy/blob/abeca76c5ca956d4bfa27e01fa4988c211ba8444/numpy/_core/include/numpy/npy_common.h#L363):

```c
typedef struct
{
    double _Val[2];
} npy_cdouble;

typedef struct
{
    float _Val[2];
} npy_cfloat;
```

It will help to use example code to describe the scalar and array layout of all of the surveyed implementation, especially the standard C `complex.h` and C++ `std::complex`.

**Reply by tigercosmos** [reply]
Thanks for correcting me, I didn't realize that there are two formats for FFTW.

THe C complex definition is as:
```
 struct _C_double_complex {
 
     double _Val[2];
 
 };
 ```
 
and the std::complex definition is:
 ```
 template<typename _Tp>
struct complex
{
// ...
private:
_Tp _M_real;
_Tp _M_imag;
};
 ```
 
 I think I got the enough information, so I will start working on it.

**Comment by tigercosmos** [comment]
@yungyuc  I guess we can close this?

**Reply by yungyuc** [reply]
We do not need to close a discussion.

**Reply by tigercosmos** [reply]
@yungyuc I think discussion is for unresolved topics? I wonder why don't we archieve the discussion if it is decided or done? The record is still there.

**Reply by yungyuc** [reply]
Discussions are not only for unresolved topics.  Issues, on the other hand, may be marked as done or rejected.

Discussions are for generating ideas.  They need guidance, but I do not want to place strong control on how we generate ideas.
